odc/stats/plugins/l34_utils/l4_water.py

```diff
@@ -3,45 +3,69 @@
 NODATA = 255
 
 
-def water_classification(xx, intertidal_mask, water_persistence):
+def water_classification(xx, water_persistence):
 
+    # Replace nan with nodata
     l4 = expr_eval(
-        "where(((a==223)|(a==221))&(b==1), 101, a)",
-        {"a": xx.classes_l3_l4.data, "b": water_persistence},
+        "where((a==a), a, nodata)",
+        {"a": xx.level_3_4.data},
         name="mark_water",
         dtype="uint8",
+        **{"nodata": NODATA},
+    )
+
+    intertidal_mask = expr_eval(
+        "where(a==_u, 1, 0)",
+        {"a": l4},
+        name="mask_intertidal",
+        dtype="uint8",
+        **{"_u": 223},
     )
 
     l4 = expr_eval(
-        "where(((a==223)|(a==221))&(b==7), 102, a)",
-        {"a": l4, "b": water_persistence},
+        "where((a==223)|(a==221), 98, a)", {"a": l4}, name="mark_water", dtype="uint8"
+    )
+
+    l4 = expr_eval(
+        "where((a==98)&(b!=1), 99, a)",
+        {"a": l4, "b": intertidal_mask},
```

Comment:
`(b!=1)` -> `~(b==_u)` to include nodata=nan case where `_u` is the value in intertidal mask above, `b` is `xx.level_3_4.data`

Reply:
Addressed accordingly.